### ISSUE TITLE
📥 Some thoughts about making it cooler to work with the total space of type family

### ISSUE THREAD

**Issue by mmcqd** [RedPRL/cooltt#348]
## Context
The main motivation (as I understand it) for introducing the ability to treat a type family as its total space was to bundle up parameters/indices that would normally have to be quantified over, making it easier to work with heavily parametrized/indexed types without having to implement implicit arguments and unification. cooltt is now at a point where the _type_ of a structure can be conveniently written down using total space conversion and record patching!
```
def category : type :=
  sig (ob : type)
      (hom : sig (s : ob) (t : ob) -> type)
      (idn : (x : ob) -> hom # [ s .= x | t .= x ])
      (seq : (f : hom) -> (g : hom # [ s .= f.t ]) -> hom # [ s .= f.s | t .= g.t ])
      (seqL : (f : hom) -> path {hom # [ s .= f.s | t .= f.t ]} {seq {idn {f.s}} f} f)
      (seqR : (f : hom) -> path {hom # [ s .= f.s | t .= f.t ]} {seq f {idn {f.t}}} f)
      (seqA : (f : hom) -> (g : hom # [ s .= f.t ]) -> (h : hom # [ s .= g.t ]) -> path {hom # [ s .= f.s | t .= h.t ]} {seq f {seq g h}} {seq {seq f g} h})
```
This is (to me, at least) just as easy to write and read as the equivalent definition using implicit arguments. It might even be easier, actually. Annoyingly, actually inhabiting this type is _not_ as nice as inhabiting the version using implicit arguments. Currently the definition of the category of cooltt types looks like this:
```
def types : category :=
  struct (ob : type)
         (hom : args => {args.s} -> {args.t})
         (idn : x => struct (fib : x => x))
         (seq : f g => struct (fib : x => g.fib {f.fib x}))
         (seqL : f i => f)
         (seqR : f i => f)
         (seqA : f g h i => struct (fib : x => {h.fib} {{g.fib} {{f.fib} x}}))
```
We're forced to construct these one element structs and project out their `fib` when we want to use them as actual `hom`s and not as the total space of a `hom`. Ideally, it would be possible to write this:
```
def types : category :=
  struct (ob : type)
         (hom : args => {args.s} -> {args.t})
         (idn : x z => z)
         (seq : f g x => g {f x}))
         (seqL : f i => f)
         (seqR : f i => f)
         (seqA : f g h i x => h {g {f x}}))
```
And have all the `struct (fib : ...)` and `... .fib` automatically inserting during elaboration. I've been thinking about how to do this, and unfortunately it is not as straight forward as I had hoped. Here's what I've come up with so far.

## Thoughts about making total spaces cooler

* If we are checking against a total space where every field except `fib` has been patched, then inserting `struct (fib: ...)` and filling in the rest of fields is easy. This can be straightforwardly added to the `intro_implicit_connectives` tactic. 

* If we are checking against a total space that does not have every field patched, we're basically out of luck, and must write down a record. This makes sense if the `fib` field is some term that must be checked, like a lambda, but if the `fib` field is synthable, it _feels_ like we ought to be able to propagate that information out of the term to fill in the unpatched fields, but I do not see a way to actually do this. If the `fib` field synths to `fam {struct ...}`, then we could use that. But if the `fam` in question is not neutral, we pretty much end up back to needing unification to figure out the values of the unpatched fields.

* Automatically inserting `.fib` projections is slightly problematic. My feeling is that we'd want to insert them as part of an `elim_implicit_connectives` style tactic, but _not_ if we're synthing a term to be projected. The reason for this is that we need to be able to project out the other fields of total space (`s` and `t` in the case of `hom`). So given `f : hom`: `f x` becomes `{f.fib} x` but `f.s` does _not_ become `{f.fib}.s`. This means that if the value of the `fib` field is a record, we would be required to manually insert the `.fib` projection. 

* The way that checking and synthing interact is problematic. If we have `f : hom # [s .= s | t .= t]`, and we go to check it against `hom # [s .= s | t .= t]`, then all is all well. The goal becomes `hom {struct (s : s) (t : t)` and `f` becomes `f.fib`. We run the `intro_fields` tactics with `struct (fib : f.fib)` and the other fields are filled in. But (!) if we have `f : hom # [s .= s | t .= t]` and we go to check it against `hom` (no patches), we're in trouble. `f` becomes `f.fib`, but as previously mentioned, we can't implicitly insert a struct when checking against a total space that isn't completely patched, so we get a type error. But this would have worked if we had not inserted the `.fib`. So we'd need to only insert it if we know that we aren't ultimately going to check against a total space.

I realize these thoughts are not really fully formed. If this is too rambly to be useful, let me know. I would love some input on whether these problems are fundamental, or I'm just approaching this in the wrong way.

**Comment by jonsterling** [comment]
@mmcqd Great thoughts! Your "dream code" is exactly what I wish I could write. I will give some thought to the "fib" issue. One quick thought is: what if we were able to detect that `fib` is the only non-trivial field in the signature? Could that help us decide whether to auto-project or auto-introduce? (Not sure this is a good idea, but wanted to bring it into the discussion.)

**Comment by mmcqd** [comment]
@jonsterling I think that should be straightforward to detect (I wrote such a check for signatures but not signature codes, I assume it wouldn't be too hard to extend), and that would be used to implement what I mentioned in the first bullet. It seems to me that auto-projecting/introducing should be happening primarily based on whether we are checking or synthing. If we are checking against such a signature, we want to intro, and if we have synthed such a signature, we want to elim (with the annoying caveats I mentioned).
